File: buildSrc/src/main/java/wolkenschloss/pool/PoolExtension.java
package wolkenschloss.pool;

import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.services.BuildServiceRegistry;

import java.io.Serializable;

public abstract class PoolExtension implements Serializable {

    public static final String DEFAULT_RUN_FILE_NAME = "root.md5";
    public static final String DEFAULT_POOL_RUN_FILE = "pool.run";

    public void initialize(BuildServiceRegistry sharedServices, DirectoryProperty buildDirectory, DirectoryProperty runDirectory) {
        getRootImageName().convention("root.qcow2");
        getCidataImageName().convention("cidata.img");
        getName().convention("testbed");

        getPoolOperations().set(sharedServices.registerIfAbsent(
                "poolops",
                PoolOperations.class,
                spec -> spec.getParameters().getPoolName().set(getName())));

        getPoolDirectory().set(buildDirectory.dir("pool"));
        getRootImageMd5File().set(runDirectory.file(DEFAULT_RUN_FILE_NAME));
        getPoolRunFile().set(runDirectory.file(DEFAULT_POOL_RUN_FILE));
    }

    public abstract Property<String> getName();

    public abstract Property<String> getRootImageName();

    public abstract Property<String> getCidataImageName();

    public abstract Property<PoolOperations> getPoolOperations();

    abstract public DirectoryProperty getPoolDirectory();

    abstract public RegularFileProperty getRootImageMd5File();

    abstract public RegularFileProperty getPoolRunFile();
}

File: buildSrc/src/main/kotlin/wolkenschloss/gradle/testbed/pool/PoolExtension.kt
package wolkenschloss.gradle.testbed.pool

import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.ProjectLayout
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import java.io.Serializable
import javax.inject.Inject

@Suppress("CdiInjectionPointsInspection")
abstract class PoolExtension @Inject constructor(private val layout: ProjectLayout) : Serializable {
    fun initialize(runDirectory: DirectoryProperty) {
        rootImageName.convention(DEFAULT_ROOT_IMAGE_NAME)
        cidataImageName.convention(DEFAULT_CIDATA_IMAGE_NAME)
        name.convention("testbed")
        poolDirectory.set(layout.buildDirectory.dir("pool"))
        rootImageMd5File.set(runDirectory.file(DEFAULT_RUN_FILE_NAME))
        poolRunFile.set(runDirectory.file(DEFAULT_POOL_RUN_FILE))
    }

    abstract val name: Property<String>
    abstract val rootImageName: Property<String>
    abstract val cidataImageName: Property<String>
    abstract val poolDirectory: DirectoryProperty
    abstract val rootImageMd5File: RegularFileProperty
    abstract val poolRunFile: RegularFileProperty

    companion object {
        const val DEFAULT_CIDATA_IMAGE_NAME = "cidata.img"
        const val DEFAULT_ROOT_IMAGE_NAME = "root.qcow2"
        const val DEFAULT_RUN_FILE_NAME = "root.md5"
        const val DEFAULT_POOL_RUN_FILE = "pool.run"

    }
}